codegenerator/cli/templates/static/codegen/src/Prometheus.res

```diff
@@ -78,32 +71,72 @@ module SafeGauge: {
       }
     | _ => Error("Label schema must be an object")
     }
+}
+
+let metricNames: Utils.Set.t<string> = Utils.Set.make()
+
+module MakeSafePromMetric = (
+  M: {
+    type t
+    let make: {"name": string, "help": string, "labelNames": array<string>} => t
+    let labels: (t, 'a) => t
+    let handleFloat: (t, float) => unit
+    let handleInt: (t, int) => unit
+  },
+): {
+  type t<'a>
+  let makeOrThrow: (~name: string, ~help: string, ~labelSchema: S.t<'a>) => t<'a>
+  let handleInt: (t<'a>, ~labels: 'a, ~value: int) => unit
+  let handleFloat: (t<'a>, ~labels: 'a, ~value: float) => unit
+} => {
+  type t<'a> = {metric: M.t, labelSchema: S.t<'a>}
 
   let makeOrThrow = (~name, ~help, ~labelSchema: S.t<'a>): t<'a> =>
-    switch labelSchema->getLabelNames {
+    switch labelSchema->Labels.getLabelNames {
     | Ok(labelNames) =>
-      PromClient.Gauge.makeGauge({
-        "name": name,
-        "help": help,
-        "labelNames": labelNames,
-      })
+      if metricNames->Utils.Set.has(name) {
+        Js.Exn.raiseError("Duplicate prometheus metric name: " ++ name)
+      } else {
+        metricNames->Utils.Set.add(name)->ignore
+        let metric = M.make({
+          "name": name,
+          "help": help,
+          "labelNames": labelNames,
+        })
+
+        {metric, labelSchema}
+      }
 
     | Error(error) => Js.Exn.raiseError(error)
     }
 
-  let makeLabels = (self: t<'a>, ~labels: 'a): t<'a> => self->PromClient.Gauge.labels(labels)
-
-  let setFloat = (self: t<'a>, ~labels: 'a, ~value) =>
-    self
-    ->makeLabels(~labels)
-    ->PromClient.Gauge.setFloat(value)
+  let handleFloat = ({metric, labelSchema}: t<'a>, ~labels: 'a, ~value) =>
+    metric
+    ->M.labels(labels->S.serializeOrRaiseWith(labelSchema))
+    ->M.handleFloat(value)
 
-  let setInt = (self: t<'a>, ~labels: 'a, ~value) =>
-    self
-    ->makeLabels(~labels)
-    ->PromClient.Gauge.set(value)
+  let handleInt = ({metric, labelSchema}: t<'a>, ~labels: 'a, ~value) =>
+    metric
+    ->M.labels(labels->S.serializeOrRaiseWith(labelSchema))
+    ->M.handleInt(value)
 }
 
+module SafeCounter = MakeSafePromMetric({
+  type t = PromClient.Counter.counter
+  let make = PromClient.Counter.makeCounter
+  let labels = PromClient.Counter.labels
+  let handleInt = PromClient.Counter.incMany
+  let handleFloat = PromClient.Counter.incMany->Utils.magic
+})
+
+module SafeGauge = MakeSafePromMetric({
+  type t = PromClient.Gauge.gauge
+  let make = PromClient.Gauge.makeGauge
+  let labels = PromClient.Gauge.labels
+  let handleInt = PromClient.Gauge.set
+  let handleFloat = PromClient.Gauge.setFloat
+})
```

Comment:
TBH this refactor was not necessary because I didn't end up using the safe counter anyways